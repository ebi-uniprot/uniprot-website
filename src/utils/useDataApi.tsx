import React, { useState, useEffect } from 'react';
import fetchData from './fetchData';
import 'regenerator-runtime/runtime';

const useDataApi = (url: string) => {
  const [data, setData] = useState<any>({});

  const getData = async () => {
    const result = await fetchData(url);
    setData(result.data);
  };

  useEffect(() => {
    getData();
  }, [url]);

  return data;
};

export default useDataApi;
