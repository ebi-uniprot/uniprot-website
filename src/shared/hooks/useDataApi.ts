import { useEffect, useReducer, useRef } from 'react';
import axios, {
  AxiosResponse,
  AxiosError,
  AxiosRequestConfig,
  AxiosProgressEvent,
  isCancel,
} from 'axios';
import joinUrl from 'url-join';

import useMessagesDispatch from './useMessagesDispatch';

import fetchData from '../utils/fetchData';
import { addMessage } from '../../messages/state/messagesActions';
import * as logging from '../utils/logging';
import { apiPrefix } from '../config/apiUrls/apiPrefix';

import {
  MessageFormat,
  MessageLevel,
} from '../../messages/types/messagesTypes';
import { Namespace } from '../types/namespaces';
import { UserPreferenceKey } from './useLocalStorage';

export type CustomError = AxiosError<undefined | { messages?: string[] }>;

const invalidFieldMessage = /Invalid fields parameter value '(?<field>[^']*)'/;
const namespacedURL = new RegExp(
  joinUrl(
    apiPrefix,
    `(?<namespace>${Object.values(Namespace).join('|')})`,
    'search'
  )
);

// TODO: possible improvement:
// See https://developers.google.com/web/tools/workbox/modules/workbox-broadcast-update#using_broadcast_update
// Listen for a service worker message saying that there is new data to get
// fresher data from cache and replace the data within this custom hook

// Threshold under which a progress event will not be emitted
// This is in order to avoid unecessary renders
const PROGRESS_THRESHOLD = 2_000; // 2s

export type UseDataAPIState<T> = {
  loading: boolean;
  progress?: number;
  data?: T;
  status?: AxiosResponse['status'];
  statusText?: AxiosResponse['statusText'];
  headers?: AxiosResponse['headers'];
  error?: CustomError;
  redirectedTo?: string;
  url?: string | null;
};

enum ActionType {
  INIT = 'INIT',
  PROGRESS = 'PROGRESS',
  SUCCESS = 'SUCCESS',
  ERROR = 'ERROR',
}

type Action<T> =
  | { type: ActionType.INIT; url?: string }
  | { type: ActionType.PROGRESS; progress: number }
  | {
      type: ActionType.SUCCESS;
      response?: AxiosResponse<T>;
      originalURL?: string;
      progress?: 1;
    }
  | { type: ActionType.ERROR; error: CustomError };

const createReducer =
  <T>() =>
  (state: UseDataAPIState<T>, action: Action<T>): UseDataAPIState<T> => {
    // eslint-disable-next-line default-case
    switch (action.type) {
      case ActionType.INIT:
        return {
          loading: true,
          url: action.url,
        };
      case ActionType.PROGRESS:
        return {
          ...state,
          progress: action.progress,
        };
      case ActionType.SUCCESS:
        // eslint-disable-next-line no-case-declarations
        const newState: UseDataAPIState<T> = {
          ...state,
          loading: false,
          progress: action.progress,
          data: action.response?.data,
          status: action.response?.status,
          statusText: action.response?.statusText,
          headers: action.response?.headers,
        };
        if (
          action.response?.request?.responseURL &&
          action.response.request.responseURL !== action.originalURL
        ) {
          newState.redirectedTo = action.response.request.responseURL;
        } else if (
          // Issue with casing in axios-mock-adapter?
          action.response?.request?.responseUrl &&
          action.response.request.responseUrl !== action.originalURL
        ) {
          newState.redirectedTo = action.response.request.responseUrl;
        }
        return newState;
      case ActionType.ERROR:
        return {
          ...state,
          loading: false,
          progress: undefined,
          status: action.error.response?.status,
          statusText: action.error.response?.statusText,
          headers: action.error.response?.headers,
          error: action.error,
        };
    }
  };

function useDataApi<T>(
  url?: string | null,
  options?: AxiosRequestConfig
): UseDataAPIState<T> {
  const [state, dispatch] = useReducer(createReducer<T>(), { loading: !!url });
  const messagesDispatch = useMessagesDispatch();
  const optionsRef = useRef(options);

  useEffect(() => {
    if (options !== optionsRef.current) {
      optionsRef.current = options;
    }
  }, [options]);

  useEffect(() => {
    // need this variable to ensure state updates don't occur when cancelled/unmounted
    // https://github.com/facebook/react/issues/14369#issuecomment-468267798
    let didCancel = false;

    // we don't require a URL, we just don't need data anymore
    // assume succes with no data
    if (!url) {
      dispatch({ type: ActionType.SUCCESS });
      return;
    }

    dispatch({ type: ActionType.INIT, url });

    // variables to handle cancellation
    // eslint-disable-next-line import/no-named-as-default-member
    const source = axios.CancelToken.source();

    // to keep track of the last time we dispatched a progress action
    let lastProgressDate: number;
    // actual request
    fetchData<T>(url, source.token, {
      ...optionsRef.current,
      onDownloadProgress: (progressEvent: AxiosProgressEvent) => {
        optionsRef.current?.onDownloadProgress?.(progressEvent);
        if (didCancel || progressEvent.progress === undefined) {
          return;
        }
        const now = Date.now();
        // If last time we dispatched a progress action was less than threshold
        if (lastProgressDate && now - lastProgressDate < PROGRESS_THRESHOLD) {
          // skip, don't want to refresh too often
          return;
        }
        dispatch({
          type: ActionType.PROGRESS,
          progress: progressEvent.progress,
        });
        lastProgressDate = now;
      },
    }).then(
      // handle ok
      (response: AxiosResponse<T>) => {
        if (didCancel) {
          return;
        }
        dispatch({
          type: ActionType.SUCCESS,
          response,
          originalURL: url,
          progress: 1,
        });
      },
      // catch error
      (error: CustomError) => {
        // Ignore 404 because it might mean "valid request but no data"
        if (error.response?.status && error.response.status !== 404) {
          logging.error(error, { tags: { origin: 'useDataApi', url } });
        }
        if (isCancel(error) || didCancel) {
          return;
        }
        dispatch({ type: ActionType.ERROR, error });
      }
    );

    // handle unmounting of the hook

    return () => {
      source.cancel();
      didCancel = true;
    };
  }, [url, optionsRef]);

  useEffect(() => {
    // TODO: when refactoring, accept a onError callback to do this from outside
    if (state.status === 400) {
      // Just in case a field is updated in the server and is no longer valid
      try {
        for (const message of state.error?.response?.data?.messages || []) {
          const messageMatch = message.match(invalidFieldMessage);
          const invalidField = messageMatch?.groups?.field;
          if (!messageMatch) {
            continue;
          }
          const nsMatch = state.url?.match(namespacedURL);
          if (!nsMatch?.groups?.namespace) {
            continue;
          }
          const key =
            `table columns for ${nsMatch?.groups?.namespace}` as UserPreferenceKey;
          const faultyArray: string[] = JSON.parse(
            localStorage.getItem(key) as string
          );
          const noFaults = faultyArray.filter(
            (column) =>
              // Clean up any wrong type of value, and the faulty value too
              column && typeof column === 'string' && column !== invalidField
          );
          // If no valid columns then store undefined which will cause the
          // default columns to be used.
          const correctArray = JSON.stringify(
            noFaults.length ? noFaults : undefined
          );
          localStorage.setItem(key, correctArray);
          // Signals to all useLocalStorage hooks in use to rerender
          window.dispatchEvent(
            new StorageEvent('storage', {
              key,
              newValue: correctArray,
              storageArea: window.localStorage,
            })
          );
          return;
        }
      } catch {
        /**/
      }
      const messageContent =
        state.error?.response?.data?.messages?.join(',') || '400 Error';
      messagesDispatch(
        addMessage({
          id: messageContent,
          content: messageContent,
          format: MessageFormat.POP_UP,
          level: MessageLevel.FAILURE,
          displayTime: 5_000,
        })
      );
    }
  }, [messagesDispatch, state.status, state.error, state.url]);

  // when changing the URL, the state is set asynchronously, this is to set it
  // to loading synchronously to avoid using previous data
  if (state.url !== url) {
    return { loading: Boolean(url), url };
  }

  return state;
}

export default useDataApi;
