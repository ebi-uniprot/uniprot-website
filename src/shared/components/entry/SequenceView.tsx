import React, { Fragment, useState } from 'react';
import { InfoList, Sequence, ExternalLink } from 'franklin-sites';
import { Link, useHistory } from 'react-router-dom';

import UniProtKBEvidenceTag from '../../../uniprotkb/components/protein-data-views/UniProtKBEvidenceTag';
import numberView, {
  Unit,
} from '../../../uniprotkb/components/protein-data-views/NumberView';
import FreeTextView from '../../../uniprotkb/components/protein-data-views/FreeTextView';
import BlastButton from '../action-buttons/Blast';
import AlignButton from '../action-buttons/Align';

import { formatLargeNumber } from '../../utils/utils';

import useDataApi from '../../hooks/useDataApi';

import externalUrls from '../../../uniprotkb/config/externalUrls';
import apiUrls from '../../config/apiUrls';

import {
  Isoform,
  SequenceCautionComment,
  MassSpectrometryComment,
  RNAEditingComment,
  AlternativeProductsComment,
} from '../../../uniprotkb/types/commentTypes';
import { UniProtkbAPIModel } from '../../../uniprotkb/adapters/uniProtkbConverter';
import { Location, LocationToPath } from '../../../app/config/urls';
import {
  IsoformNotes,
  SequenceUIModel,
} from '../../../uniprotkb/adapters/sequenceConverter';

export type SequenceData = {
  value: string;
  length: number;
  molWeight: number;
  crc64: string;
};

type SequenceViewProps = {
  accession: string;
  data: SequenceUIModel;
};

export const SequenceInfo: React.FC<{
  isoformId: string;
  isoformSequence?: SequenceData;
  lastUpdateDate?: string | null;
  isCanonical?: boolean;
}> = ({ isoformId, isoformSequence, lastUpdateDate, isCanonical = false }) => {
  const [isoformToFetch, setIsoformToFetch] = useState<string>();

  const history = useHistory();

  const { data, loading } = useDataApi<UniProtkbAPIModel>(
    isoformToFetch && apiUrls.entry(isoformToFetch)
  );

  const dataToDisplay = data?.sequence || isoformSequence;

  const infoData = [
    {
      title: 'Length',
      content: dataToDisplay && dataToDisplay.length,
    },
    {
      title: 'Mass (Da)',
      content: dataToDisplay && formatLargeNumber(dataToDisplay.molWeight),
    },
    {
      title: 'Last updated',
      content: lastUpdateDate,
    },
    {
      title: 'Checksum',
      content: dataToDisplay && dataToDisplay.crc64,
    },
  ];

  return (
    <Sequence
      sequence={dataToDisplay && dataToDisplay.value}
      onShowSequence={() => setIsoformToFetch(isoformId)}
      infoData={infoData}
      accession={isoformId}
      downloadUrl={apiUrls.sequenceFasta(isoformId)}
      onBlastClick={() =>
        history.push(LocationToPath[Location.Blast], {
          parameters: { sequence: dataToDisplay && dataToDisplay.value },
        })
      }
      // This callbacks has been commented out as
      //  the basket is not yet implemented
      // onAddToBasketClick={() => {}}
      isCollapsible={!isCanonical}
      isLoading={loading}
    />
  );
};

export const IsoformInfo: React.FC<{
  isoformData: Isoform;
  canonicalAccession: string;
  isoformNotes?: IsoformNotes;
}> = ({ isoformData, canonicalAccession, isoformNotes }) => {
  let note;
  const regex = new RegExp(isoformData.name.value, 'gi');
  for (const key in isoformNotes) {
    if (key.match(regex)) {
      note = isoformNotes[key];
      break;
    }
  }

  const infoListData = [
    {
      title: 'Name',
      content: isoformData.name.value,
    },
    {
      title: 'Synonyms',
      content: (isoformData?.synonyms ?? []).map((syn) => syn.value).join(', '),
    },
    {
      title: 'Note',
      content: note && <FreeTextView comments={note} showMolecule={false} />,
    },
    {
      title: 'Differences from canonical',
      content: isoformData.varSeqs && !!isoformData.varSeqs.length && (
        <ul className="no-bullet">
          {isoformData.varSeqs.map(
            ({ location, alternativeSequence, evidences }) => (
              <li key={`${location.start.value}-${location.end.value}`}>
                <Link
                  to={`/blast/accession/${canonicalAccession}/positions/${location.start.value}-${location.end.value}`}
                >{`${location.start.value}-${location.end.value}: `}</Link>
                {alternativeSequence && alternativeSequence.originalSequence
                  ? `${alternativeSequence.originalSequence}  â†’ ${
                      alternativeSequence.alternativeSequences &&
                      alternativeSequence.alternativeSequences.join(', ')
                    }`
                  : 'Missing'}
                {evidences && <UniProtKBEvidenceTag evidences={evidences} />}
              </li>
            )
          )}
        </ul>
      ),
    },
    {
      title: 'Note',
      content:
        isoformData.note &&
        isoformData.note.texts.map((note) => note.value).join(', '),
    },
  ];
  // TODO isoformData.sequenceIds is used to get the features for
  // splice variants - they need to be somehow displayed
  const name = isoformData.isoformIds.join(', ');
  return (
    <Fragment key={isoformData.isoformIds.join('')}>
      <hr />
      <h3 id={name}>{name}</h3>
      {isoformData.isoformSequenceStatus === 'Displayed' && (
        <p>
          {'This isoform has been chosen as the '}
          <strong>canonical</strong>
          {' sequence. All positional information in '}
          {'this entry refers to it. This is also the sequence '}
          that appears in the downloadable versions of the entry.
        </p>
      )}
      {isoformData.isoformSequenceStatus === 'External' && (
        <section>
          <p>
            The sequence of this isoform can be found in the external entry
            linked below. Isoforms of the same protein are often annotated in
            two different entries if their sequences differ significantly.{' '}
          </p>
          {/* TODO: this is hacky and temporary until we sort out
          external isoforms */}
          <Link
            className="button secondary"
            to={`/uniprotkb/${isoformData.isoformIds[0].substring(
              0,
              isoformData.isoformIds[0].length - 2
            )}`}
          >
            View isoform
          </Link>
        </section>
      )}
      <InfoList infoData={infoListData} columns isCompact />
    </Fragment>
  );
};

export const SequenceCautionView: React.FC<{
  data: SequenceCautionComment[];
}> = ({ data }) => {
  return (
    <>
      {data.map(({ sequence, sequenceCautionType, note, evidences }) => (
        <section
          className="text-block"
          key={`${sequenceCautionType}-${sequence}`}
        >
          {`The sequence `}
          <ExternalLink url={externalUrls.ENA(sequence)}>
            {sequence}
          </ExternalLink>
          {` differs from that shown. Reason: ${sequenceCautionType} `}
          {note}
          {evidences && <UniProtKBEvidenceTag evidences={evidences} />}
        </section>
      ))}
    </>
  );
};

export const MassSpectrometryView: React.FC<{
  data: MassSpectrometryComment[];
}> = ({ data }) => (
  <>
    {data.map((item) => (
      <section className="text-block" key={`${item.molWeight}${item.method}`}>
        {item.molecule && <h3>{item.molecule}</h3>}
        {`Molecular mass is ${numberView({
          value: item.molWeight,
          unit: Unit.DA,
        })}. `}
        {item.method && `Determined by ${item.method}. `}
        {item.note}
        <UniProtKBEvidenceTag evidences={item.evidences} />
      </section>
    ))}
  </>
);

export const RNAEditingView: React.FC<{ data: RNAEditingComment[] }> = ({
  data,
}) => (
  <>
    {data.map((item) => (
      <section
        className="text-block"
        key={`${
          item.positions && item.positions.map((pos) => pos.position).join('')
        }`}
      >
        {item.positions && (
          <div>
            {'Edited at positions '}
            {item.positions.map((position) => (
              <span key={position.position}>
                {position.position}{' '}
                <UniProtKBEvidenceTag evidences={position.evidences} />
              </span>
            ))}
          </div>
        )}
        {item.note && (
          <div>
            {item.note.texts.map((text) => (
              <span key={text.value}>
                {text.value}{' '}
                {text.evidences && (
                  <UniProtKBEvidenceTag evidences={text.evidences} />
                )}
              </span>
            ))}
          </div>
        )}
      </section>
    ))}
  </>
);

export const IsoformView: React.FC<{
  alternativeProducts: AlternativeProductsComment;
  canonicalComponent?: JSX.Element;
  includeSequences?: boolean;
  canonicalAccession: string;
  isoformNotes?: IsoformNotes;
}> = ({
  alternativeProducts,
  canonicalComponent,
  includeSequences = true,
  canonicalAccession,
  isoformNotes,
}) => {
  let isoformCountNode;
  const { isoforms, events } = alternativeProducts;
  if (isoforms && events) {
    isoformCountNode = (
      <p>
        {`This entry describes `}
        <strong>{isoforms.length}</strong>
        {` isoforms produced by `}
        <strong>{events.join(' & ')}</strong>.
      </p>
    );
  }

  let notesNode;
  const texts = alternativeProducts.note?.texts;
  if (texts) {
    notesNode = <p>{texts.map((text) => text.value).join(' ')}</p>;
  }

  let isoformsNode;
  if (isoforms) {
    isoformsNode = isoforms.map((isoform) => {
      const isoformComponent = (
        <SequenceInfo isoformId={isoform.isoformIds[0]} />
      );
      return (
        <Fragment key={isoform.isoformIds.join('')}>
          <IsoformInfo
            isoformData={isoform}
            canonicalAccession={canonicalAccession}
            isoformNotes={isoformNotes}
          />
          {includeSequences && isoform.isoformSequenceStatus !== 'External' && (
            <>
              {canonicalComponent &&
              isoform.isoformSequenceStatus === 'Displayed'
                ? canonicalComponent
                : isoformComponent}
            </>
          )}
        </Fragment>
      );
    });
  }
  return (
    <>
      {isoformCountNode}
      {notesNode}
      {isoformsNode}
    </>
  );
};

const SequenceView: React.FC<SequenceViewProps> = ({ accession, data }) => {
  const sequenceInfoData = [
    {
      title: 'Sequence status',
      content: data.status,
    },
    {
      title: 'Sequence processing',
      content: data.processing,
    },
  ];

  // Every entry should have a sequence
  if (!data.sequence) {
    return null;
  }

  const canonicalComponent = (
    <SequenceInfo
      isoformId={accession}
      isoformSequence={data.sequence}
      lastUpdateDate={data.lastUpdateDate}
      isCanonical
    />
  );

  if (!data.alternativeProducts && data.sequence) {
    return canonicalComponent;
  }

  if (!data.alternativeProducts) {
    return null;
  }

  const allIsoformIds = data.alternativeProducts.isoforms
    .map((isoform) => isoform.isoformIds)
    .flat();

  return (
    <>
      <div className="button-group">
        <BlastButton
          selectedEntries={allIsoformIds}
          textSuffix={`${allIsoformIds.length} isoform${
            allIsoformIds.length === 1 ? '' : 's'
          }`}
        />
        <AlignButton
          selectedEntries={allIsoformIds}
          textSuffix={
            allIsoformIds.length === 1
              ? undefined
              : `${allIsoformIds.length} isoforms`
          }
        />
        {/* Missing Add to basket */}
      </div>

      <InfoList infoData={sequenceInfoData} columns />
      <IsoformView
        alternativeProducts={data.alternativeProducts}
        canonicalComponent={canonicalComponent}
        canonicalAccession={accession}
        isoformNotes={data.isoformNotes}
      />
    </>
  );
};

export default SequenceView;
