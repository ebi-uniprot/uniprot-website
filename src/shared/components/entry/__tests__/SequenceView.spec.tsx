import React from 'react';
import { createMemoryHistory } from 'history';

import { fireEvent } from '@testing-library/react';
import renderWithRedux from '../../../../shared/__test-helpers__/RenderWithRedux';

import SequenceView from '../SequenceView';

import SequenceUIDataJson from './__mocks__/sequenceUIData.json';

jest.mock('../../../hooks/useDataApi.ts');
import useDataApi from '../../../hooks/useDataApi';

useDataApi.mockImplementation((url) =>
  url
    ? {
        data: {
          sequence: {
            length: 100,
            molWeight: 100000,
            crc64: 'ABCSSDDD',
            value: 'MNOPQRSTUVWXYZ',
          },
        },
      }
    : {}
);

let rendered;
const history = createMemoryHistory();

describe('SequenceView component', () => {
  beforeEach(() => {
    window.SVGElement.prototype.getBBox = () => ({
      width: 10,
      height: 10,
    });
    rendered = renderWithRedux(
      <SequenceView data={SequenceUIDataJson} accession="P05067" />,
      { history }
    );
  });

  afterEach(() => {
    delete window.SVGElement.prototype.getBBox;
  });

  test('should render SequenceInfo with provided sequence info (canonical)', () => {
    const { asFragment } = rendered;
    expect(asFragment()).toMatchSnapshot();
  });

  test('should trigger Isoform sequence load', async () => {
    const { getByText, queryByText, findByText } = rendered;
    const viewSequenceButton = getByText(/Show sequence/i);
    expect(queryByText(/ABCSSDDD/)).toBeNull();
    fireEvent.click(viewSequenceButton);
    const sequence = await findByText(/ABCSSDDD/);
    expect(sequence).toBeTruthy();
  });

  test('should submit Blast', async () => {
    const { getByText, findByText } = rendered;
    const toolsButton = getByText(/Tools/i);
    fireEvent.click(toolsButton);
    const blastButton = await findByText('BLAST');
    fireEvent.click(blastButton);
    expect(history.location.pathname).toBe('/blast');
    expect(history.location.state.parameters).toStrictEqual({
      sequence: 'ATGHASOADSDKDSAJALDSAMCMnDSKMDSKNCSAKN',
    });
  });
});
