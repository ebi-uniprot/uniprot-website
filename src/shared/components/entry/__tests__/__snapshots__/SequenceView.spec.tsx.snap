// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SequenceView component should render SequenceInfo with provided sequence info (canonical) 1`] = `
<DocumentFragment>
  <div
    class="button-group"
  >
    <button
      class="button tertiary"
      title="Run 2 BLAST jobs against the selected entries"
      type="button"
    >
      BLAST 2 isoforms
    </button>
    <button
      class="button tertiary"
      title="Run an Align job against 2 entries"
      type="button"
    >
      Align 2 isoforms
    </button>
  </div>
  <div
    class="info-list info-list--columns"
  >
    <div
      class="info-list__item"
    >
      <div
        class="info-list__title"
      >
        <h5
          class="bold"
        >
          Sequence status
        </h5>
      </div>
      <div
        class="info-list__content"
      >
        Sequence status
      </div>
    </div>
    <div
      class="info-list__item"
    >
      <div
        class="info-list__title"
      >
        <h5
          class="bold"
        >
          Sequence processing
        </h5>
      </div>
      <div
        class="info-list__content"
      >
        Sequence processing
      </div>
    </div>
  </div>
  <p>
    This entry describes 
    <strong>
      2
    </strong>
     isoforms produced by 
    <strong>
      event 1 & event 2
    </strong>
    .
  </p>
  <p>
    some text
  </p>
  <hr />
  <h3
    id="P05067-1"
  >
    P05067-1
  </h3>
  <p>
    This isoform has been chosen as the 
    <strong>
      canonical
    </strong>
     sequence. All positional information in this entry refers to it. This is also the sequence that appears in the downloadable versions of the entry.
  </p>
  <div
    class="info-list info-list--columns info-list--compact"
  >
    <div
      class="info-list__item"
    >
      <div
        class="info-list__title"
      >
        <h5
          class="bold"
        >
          Name
        </h5>
      </div>
      <div
        class="info-list__content"
      >
        APP770
      </div>
    </div>
    <div
      class="info-list__item"
    >
      <div
        class="info-list__title"
      >
        <h5
          class="bold"
        >
          Synonyms
        </h5>
      </div>
      <div
        class="info-list__content"
      >
        PreA4 770
      </div>
    </div>
    <div
      class="info-list__item"
    >
      <div
        class="info-list__title"
      >
        <h5
          class="bold"
        >
          Note
        </h5>
      </div>
      <div
        class="info-list__content"
      >
        A major isoform.
      </div>
    </div>
  </div>
  <section
    class="sequence-container"
  >
    <div
      class="action-bar button-group"
    >
      <div
        class="dropdown-container"
      >
        <button
          class="button dropdown tertiary"
          type="button"
        >
          Tools
        </button>
        <div
          class="dropdown-menu"
        >
          <ul
            class="no-bullet"
          >
            <li>
              <button
                class="button tertiary"
                type="button"
              >
                BLAST
              </button>
            </li>
            <li>
              <a
                href="//web.expasy.org/cgi-bin//protparam/protparam?P05067"
                rel="noopener noreferrer"
                target="_blank"
              >
                ProtParam
              </a>
            </li>
            <li>
              <a
                href="//web.expasy.org/cgi-bin//protscale/protscale.pl?P05067"
                rel="noopener noreferrer"
                target="_blank"
              >
                ProtScale
              </a>
            </li>
            <li>
              <a
                href="//web.expasy.org/cgi-bin//compute_pi/pi_tool?P05067"
                rel="noopener noreferrer"
                target="_blank"
              >
                Compute pI/Mw
              </a>
            </li>
            <li>
              <a
                href="//web.expasy.org/cgi-bin//peptide_mass/peptide-mass.pl?P05067"
                rel="noopener noreferrer"
                target="_blank"
              >
                PeptideMass
              </a>
            </li>
            <li>
              <a
                href="//web.expasy.org/cgi-bin//peptide_cutter/peptidecutter.pl?P05067"
                rel="noopener noreferrer"
                target="_blank"
              >
                PeptideCutter
              </a>
            </li>
          </ul>
        </div>
      </div>
      <a
        class="button tertiary"
        download=""
        href="https://wwwdev.ebi.ac.uk/uniprot/api/uniprotkb/accession/P05067.fasta"
      >
        <svg
          fill="currentColor"
          viewBox="0 0 32 32"
        >
          <path
            d="M27.6 21h3.9v7.7H.5v-7.8h3.9v4h23.2v-4zm-3.9-9.7L16 22.9 8.2 11.3h5.9V1.6h3.8v9.7h5.9z"
          />
        </svg>
        Download
      </a>
      <div
        class="dropdown-container"
      >
        <button
          class="button dropdown tertiary"
          type="button"
        >
          Highlight
        </button>
        <div
          class="dropdown-menu"
        >
          <div
            class="dropdown-menu__content"
          >
            <label
              for="P05067-Polar"
            >
              <input
                data-testid="sequence-highlight-checkbox"
                id="P05067-Polar"
                type="checkbox"
              />
              Polar
            </label>
            <label
              for="P05067-Hydrophobic"
            >
              <input
                data-testid="sequence-highlight-checkbox"
                id="P05067-Hydrophobic"
                type="checkbox"
              />
              Hydrophobic
            </label>
          </div>
        </div>
      </div>
      <button
        class="button tertiary"
        type="button"
      >
        Copy FASTA
      </button>
    </div>
    <div
      class="info-list info-list--columns info-list--compact"
    >
      <div
        class="info-list__item"
      >
        <div
          class="info-list__title"
        >
          <h5
            class="bold"
          >
            Length
          </h5>
        </div>
        <div
          class="info-list__content"
        >
          1234
        </div>
      </div>
      <div
        class="info-list__item"
      >
        <div
          class="info-list__title"
        >
          <h5
            class="bold"
          >
            Mass (Da)
          </h5>
        </div>
        <div
          class="info-list__content"
        >
          5,678
        </div>
      </div>
      <div
        class="info-list__item"
      >
        <div
          class="info-list__title"
        >
          <h5
            class="bold"
          >
            Last updated
          </h5>
        </div>
        <div
          class="info-list__content"
        >
          12/04/2010
        </div>
      </div>
      <div
        class="info-list__item"
      >
        <div
          class="info-list__title"
        >
          <h5
            class="bold"
          >
            Checksum
          </h5>
        </div>
        <div
          class="info-list__content"
        >
          ASSDFFFDSAASDDSDFSASDF
        </div>
      </div>
    </div>
    <div
      class="sequence"
    >
      <div
        class="sequence__sequence"
      >
        <span
          class="sequence__sequence__chunk"
        >
          <svg
            height="30"
            width="100"
          >
            <text
              class="sequence__sequence__chunk__counter"
              x="100"
              y="10"
            >
              10
            </text>
            <text
              class="sequence__sequence__chunk__sequence"
              x="0"
              y="20"
            >
              ATGHASOADS
            </text>
          </svg>
        </span>
        <span
          class="sequence__sequence__chunk"
        >
          <svg
            height="30"
            width="100"
          >
            <text
              class="sequence__sequence__chunk__counter"
              x="100"
              y="10"
            >
              20
            </text>
            <text
              class="sequence__sequence__chunk__sequence"
              x="0"
              y="20"
            >
              DKDSAJALDS
            </text>
          </svg>
        </span>
        <span
          class="sequence__sequence__chunk"
        >
          <svg
            height="30"
            width="100"
          >
            <text
              class="sequence__sequence__chunk__counter"
              x="100"
              y="10"
            >
              30
            </text>
            <text
              class="sequence__sequence__chunk__sequence"
              x="0"
              y="20"
            >
              AMCMnDSKMD
            </text>
          </svg>
        </span>
        <span
          class="sequence__sequence__chunk"
        >
          <svg
            height="30"
            width="100"
          >
            <text
              class="sequence__sequence__chunk__counter"
              x="100"
              y="10"
            />
            <text
              class="sequence__sequence__chunk__sequence"
              x="0"
              y="20"
            >
              SKNCSAKN
            </text>
          </svg>
        </span>
      </div>
    </div>
  </section>
  <hr />
  <h3
    id="P05067-2"
  >
    P05067-2
  </h3>
  <div
    class="info-list info-list--columns info-list--compact"
  >
    <div
      class="info-list__item"
    >
      <div
        class="info-list__title"
      >
        <h5
          class="bold"
        >
          Name
        </h5>
      </div>
      <div
        class="info-list__content"
      >
        APP770-2
      </div>
    </div>
    <div
      class="info-list__item"
    >
      <div
        class="info-list__title"
      >
        <h5
          class="bold"
        >
          Synonyms
        </h5>
      </div>
      <div
        class="info-list__content"
      >
        PreA4 770-2
      </div>
    </div>
    <div
      class="info-list__item"
    >
      <div
        class="info-list__title"
      >
        <h5
          class="bold"
        >
          Note
        </h5>
      </div>
      <div
        class="info-list__content"
      >
        A not so major isoform.
      </div>
    </div>
  </div>
  <button
    class="button secondary"
    type="button"
  >
    Show sequence 
  </button>
</DocumentFragment>
`;
